src/platform_linux/shmem.c

```diff
@@ -0,0 +1,1152 @@
+ */
+typedef struct shm_frag_info {
+   void *shm_frag_addr;                // Start address of this memory fragment
+                                       // NULL => tracking fragment is empty
+   size_t   shm_frag_size;             // bytes
+   int      shm_frag_allocated_to_pid; // Allocated to this OS-pid
+   threadid shm_frag_allocated_to_tid; // Allocated to this Splinter thread-ID
+   int      shm_frag_freed_by_pid;     // OS-pid that freed this
+   threadid shm_frag_freed_by_tid;     // Splinter thread-ID that freed this
+} shm_frag_info;
+
+/*
+ * All memory allocations of this size or larger will be tracked in the
+ * above fragment tracker array. For large inserts workload, we allocate large
+ * memory chunks for fingerprint array, which is more than a MiB. For scans,
+ * splinterdb_iterator_init() allocates memory for an iterator which is ~92+KiB.
+ * Set this to a lower value so we can re-cycle free fragments for iterators
+ * also.
+ */
+#if SPLINTER_DEBUG
+#   define SHM_LARGE_FRAG_SIZE (90 * KiB)
+#else
+#   define SHM_LARGE_FRAG_SIZE (38 * KiB)
+#endif // SPLINTER_DEBUG
+
+/*
+ * In the worst case we may have all threads performing activities that need
+ * such large memory fragments. We track up to twice the # of configured
+ * threads, which is still a small array to search.
+ */
+#define SHM_NUM_LARGE_FRAGS (MAX_THREADS * 2)
+
+/*
+ * -----------------------------------------------------------------------------
+ * shmem_info{}: Shared memory Control Block:
+ *
+ * Core structure describing shared memory segment created. This lives right
+ * at the start of the allocated shared segment.
+ *
+ * NOTE(s):
+ *  - shm_large_frag_hip tracks the highest-address of all the large fragments
+ *    that are tracked. This is an optimization to short-circuit the search
+ *    done when freeing any fragment, to see if it's a large-fragment.
+ *  - shm_nf_search_skipped counts the # of times we were able to short-circuit
+ *    this search using above optimization.
+ * -----------------------------------------------------------------------------
+ */
+typedef struct shmem_info {
+   void *shm_start; // Points to start address of shared segment.
+   void *shm_end;   // Points to end address; one past end of sh segment
+   void *shm_next;  // Points to next 'free' address to allocate from.
+   void *shm_splinterdb_handle;
+   void *shm_large_frag_hip; // Highest addr of large-fragments tracked
+
+   platform_spinlock shm_mem_frags_lock;
+   // Protected by shm_mem_frags_lock. Must hold to read or modify.
+   shm_frag_info shm_mem_frags[SHM_NUM_LARGE_FRAGS];
+   uint32        shm_num_frags_tracked;
+   uint32        shm_num_frags_inuse;
+   uint32        shm_num_frags_inuse_HWM;
+   size_t        shm_used_by_large_frags; // Actually reserved
+   size_t        shm_nfrees;              // # of calls to free memory
+   size_t        shm_nf_search_skipped;   // See above;
```

Comment:
Discussed... Rob prefers that this come out. Will do so ...

Reply:
Took it out as suggested. (Upon a re-visit a few months since we spoke, I do think this kind of monitoring counter will be beneficial during perf benchmarking.)